sdk/nodejs/tests/runtime/props.spec.ts

```diff
@@ -95,18 +143,111 @@ describe("runtime", () => {
                 "secret1": secret(1),
                 "secret2": secret(undefined),
             };
+
             // Serialize and then deserialize all the properties, checking that they round-trip as expected.
-            const transfer = gstruct.Struct.fromJavaScript(
+            runtime._setFeatureSupport("secrets", true);
+            let transfer = gstruct.Struct.fromJavaScript(
                 await runtime.serializeProperties("test", inputs));
-            const result = runtime.deserializeProperties(transfer);
+            let result = runtime.deserializeProperties(transfer);
+            assert.ok(runtime.isRpcSecret(result.secret1));
+            assert.ok(runtime.isRpcSecret(result.secret2));
+            assert.strictEqual(runtime.unwrapRpcSecret(result.secret1), 1);
+            assert.strictEqual(runtime.unwrapRpcSecret(result.secret2), null);
+
+            // Serialize and then deserialize all the properties, checking that they round-trip as expected.
+            runtime._setFeatureSupport("secrets", false);
+            transfer = gstruct.Struct.fromJavaScript(
+                await runtime.serializeProperties("test", inputs));
+            result = runtime.deserializeProperties(transfer);
             assert.strictEqual(result.secret1, 1);
             assert.strictEqual(result.secret2, undefined);
+
             runtime._setTestModeEnabled(false);
         }));
+        it("marshals resource references correctly during preview", asyncTest(async () => {
+            runtime._setTestModeEnabled(false);
+            runtime._setIsDryRun(true);
+            runtime.setMocks(new TestMocks());
+
+            const component = new TestComponentResource("test");
+            const custom = new TestCustomResource("test");
+
+            const componentURN = await component.urn.promise();
+            const customURN = await custom.urn.promise();
+            const customID = await custom.id.promise();
+
+            const inputs: Inputs = {
+                "component": component,
+                "custom": custom,
+            };
+
+            runtime._setFeatureSupport("resourceReferences", true);
+
+            let serialized = await runtime.serializeProperties("test", inputs);
+            assert.deepEqual(serialized, {
+                "component": {
+                    [runtime.specialSigKey]: runtime.specialResourceSig,
+                    "urn": componentURN,
+                },
+                "custom": {
+                    [runtime.specialSigKey]: runtime.specialResourceSig,
+                    "urn": customURN,
+                    "id": customID,
+                },
+            });
+
+            runtime._setFeatureSupport("resourceReferences", false);
+            serialized = await runtime.serializeProperties("test", inputs);
+            assert.deepEqual(serialized, {
+                "component": componentURN,
+                "custom": customID ? customID : runtime.unknownValue,
```

Comment:
I'm a bit confused by this part. Shouldn't the ID be unknown in the preview? When does this expression evaluate to true?

Reply:
I believe it would be known if the user supplied the name rather than using auto-naming.